File: js/algorithm/bst/bst_node.js
const NOT_FOUND = -1

class BSTNode {
  constructor(data) {
    this.value = data
    this.left = null
    this.right = null
  }

  static _insertHelper(val, node) {
    if (node === null) return new BSTNode(val)

    if (val < node.value)
      node.left = BSTNode._insertHelper(val, node.left)
    else
      node.right = BSTNode._insertHelper(val, node.right)
    return node
  }

  static _removeHelper(val, node) {
    if (node === null) return null

    if (val < node.value) {
      node.left = BSTNode._removeHelper(val, node.left)
    } else if (node.value < val) {
      node.right = BSTNode._removeHelper(val, node.right)
    } else {
      if ((node.left === null) && (node.right === null))
        return null
      else if (node.left === null)
        result = node.right
      else if (node.right === null)
        result = node.left
      else {
        let successor = node.right.findMin()
        node.value = successor
        node.right = BSTNode._removeHelper(successor, node.right)
      }
    }
    return node
  }

  search(val) {
    if (this.value === val)
      return true
    if (this.value > val)
      return this.left === null ? false : this.left.search(val)
    else
      return this.right === null ? false : this.right.search(val)
  }

  findMin() {
    return this.left === null ? this.value : this.left.findMin()
  }

  findMax() {
    return this.right === null ? this.value : this.right.findMax()
  }

  findPredecessor(val) {
    let predecessor = NOT_FOUND
    let node = this
    while ((node !== null) && (node.value !== val)) {
      if (node.value < val) {
        predecessor = node.value
        node = node.right
      } else {
        node = node.left
      }
    }

    if (node === null)
      return NOT_FOUND
    if (node.left !== null)
      return node.left.findMax()
    else
      return predecessor
  }

  findSuccessor(val) {
    let successor = NOT_FOUND
    let node = this
    while ((node !== null) && (node.value !== val)) {
      if (node.value > val) {
        successor = node.value
        node = node.left
      } else {
        node = node.right
      }
    }

    if (node === null)
      return NOT_FOUND
    if (node.right !== null)
      return node.right.findMin()
    else
      return successor
  }

  insert(val) {
    return BSTNode._insertHelper(val, this)
  }

  remove(val) {
    return BSTNode._removeHelper(val, this)
  }

  inorder() {
    let result = []
    if (this.left !== null)
      result = result.concat(this.left.inorder())
    result.push(this.value)
    if (this.right !== null)
      result = result.concat(this.right.inorder())
    return result
  }
}

module.exports = BSTNode

File: python/algorithm/bst/bstnode.py

from http.client import NOT_FOUND

NOT_FOUND = -1

class BSTNode:
    def __init__(self, data):
        self.value = data
        self.left = None
        self.right = None

    @classmethod
    def insert_helper(cls, val, node):
        if node == None:
            return BSTNode(val)

        if val < node.value:
            node.left = BSTNode.insert_helper(val, node.left)
        else:
            node.right = BSTNode.insert_helper(val, node.right)
        return node

    @classmethod
    def remove_helper(cls, val, node):
        if node == None:
            return None

        if val < node.value:
            node.left = BSTNode.remove_helper(val, node.left)
        elif node.value < val:
            node.right = BSTNode.remove_helper(val, node.right)
        else:
            if node.left == None and node.right == None:
                return None
            elif node.left == None:
                node = node.right
            elif node.right == None:
                node = node.left
            else:
                successor = node.right.find_min()
                node.value = successor
                node.right = BSTNode.remove_helper(successor, node.right)
        return node

    def search(self, val):
        if self.value == val:
            return True
        if val < self.value:
            return False if self.left == None else self.left.search(val)
        else:
            return False if self.right == None else self.right.search(val)

    def find_min(self):
        return self.value if self.left == None else self.left.find_min()

    def find_max(self):
        return self.value if self.right == None else self.right.find_max()

    def find_predecessor(self, val):
        predecessor = NOT_FOUND
        node = self
        while node != None and node.value != val:
            if node.value < val:
                predecessor = node.value
                node = node.right
            else:
                node = node.left

        if node == None:
            return NOT_FOUND
        if node.left != None:
            return node.left.find_max()
        else:
            return predecessor

    def find_successor(self, val):
        successor = NOT_FOUND
        node = self
        while node != None and node.value != val:
            if val < node.value:
                successor = node.value
                node = node.left
            else:
                node = node.right

        if node == None:
            return NOT_FOUND
        if node.right != None:
            return node.right.find_min()
        else:
            return successor

    def insert(self, val):
        return BSTNode.insert_helper(val, self)

    def remove(self, val):
        return BSTNode.remove_helper(val, self)

    def inorder(self):
        result = []
        if self.left != None:
            result.extend(self.left.inorder())
        result.append(self.value)
        if self.right != None:
            result.extend(self.right.inorder())
        return result
